File: pkg/pagination/pagination.go
package pagination

import (
	"strings"

	"gorm.io/gorm"
)

const (
	DefaultPage     = 1
	DefaultPageSize = 20
	MaxPageSize     = 100
)

type PageRequest struct {
	Page      int    `json:"page"`
	PageSize  int    `json:"page_size"`
	SortBy    string `json:"sort_by,omitempty"`
	SortOrder string `json:"sort_order,omitempty"`
	Keyword   string `json:"keyword,omitempty"`
}

type SortField struct {
	Field string
	Order string
}

func NewPageRequest(page, pageSize int) *PageRequest {
	if page <= 0 {
		page = DefaultPage
	}
	if pageSize <= 0 {
		pageSize = DefaultPageSize
	}
	if pageSize > MaxPageSize {
		pageSize = MaxPageSize
	}
	return &PageRequest{
		Page:     page,
		PageSize: pageSize,
	}
}

func (p *PageRequest) Offset() int {
	return (p.Page - 1) * p.PageSize
}

func (p *PageRequest) Limit() int {
	return p.PageSize
}

type PageResult[T any] struct {
	Items    []T   `json:"items"`
	Total    int64 `json:"total"`
	Page     int   `json:"page"`
	PageSize int   `json:"page_size"`
}

func NewPageResult[T any](items []T, total int64, page, pageSize int) *PageResult[T] {
	return &PageResult[T]{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
	}
}

func ApplySort(query *gorm.DB, sortBy, sortOrder string, allowedFields map[string]string) *gorm.DB {
	if sortBy == "" {
		return query
	}
	column, ok := allowedFields[sortBy]
	if !ok {
		return query
	}
	order := "ASC"
	if strings.EqualFold(sortOrder, "desc") {
		order = "DESC"
	}
	return query.Order(column + " " + order)
}

func ApplyPagination(query *gorm.DB, page, pageSize int) *gorm.DB {
	if page <= 0 {
		page = DefaultPage
	}
	if pageSize <= 0 {
		pageSize = DefaultPageSize
	}
	if pageSize > MaxPageSize {
		pageSize = MaxPageSize
	}
	return query.Offset((page - 1) * pageSize).Limit(pageSize)
}

func ApplyKeyword(query *gorm.DB, keyword string, fields ...string) *gorm.DB {
	if keyword == "" || len(fields) == 0 {
		return query
	}
	kw := "%" + keyword + "%"
	expr := query.Where(fields[0]+" LIKE ?", kw)
	for _, f := range fields[1:] {
		expr = expr.Or(f + " LIKE ?", kw)
	}
	return query.Where(expr)
}

type Pagination struct {
	Page     int `json:"page"`
	PageSize int `json:"page_size"`
	Total    int `json:"total"`
}

type PaginatedResult struct {
	Data       interface{} `json:"data"`
	Pagination Pagination  `json:"pagination"`
}

func NewPagination(page, pageSize int) *Pagination {
	if page <= 0 {
		page = DefaultPage
	}
	if pageSize <= 0 {
		pageSize = DefaultPageSize
	}
	if pageSize > MaxPageSize {
		pageSize = MaxPageSize
	}
	return &Pagination{
		Page:     page,
		PageSize: pageSize,
	}
}

func (p *Pagination) Offset() int {
	return (p.Page - 1) * p.PageSize
}

func (p *Pagination) Limit() int {
	return p.PageSize
}

func (p *Pagination) SetTotal(total int) {
	p.Total = total
}

func NewPaginatedResult(data interface{}, page, pageSize, total int) *PaginatedResult {
	pg := NewPagination(page, pageSize)
	pg.SetTotal(total)
	return &PaginatedResult{
		Data:       data,
		Pagination: *pg,
	}
}

File: web-admin/src/types/pagination.ts
/**
 * 统一分页类型。
 *
 * 后端不同接口存在三种分页返回包裹（历史原因）：
 *   1. { data: T[], pagination: { page, page_size, total } }
 *   2. { list: T[], total }
 *   3. { items: T[], page, page_size, total }
 * 本文件统一对外类型，并通过 toPageResult 在列表层归一化，供 useTableList 消费。
 */

/** 分页请求参数（后端统一 page / page_size） */
export interface PageQuery {
  page: number
  page_size: number
  /** 扩展过滤字段，由具体列表接口自行扩展 */
  [key: string]: unknown
}

/** 后端嵌套分页元数据 */
export interface PaginationMeta {
  page: number
  page_size: number
  total: number
  total_pages?: number
}

/** 后端嵌套分页响应：{ data, pagination } */
export interface PaginatedResult<T> {
  data: T[]
  pagination: PaginationMeta
}

/** 前端统一列表形态：{ list, total }（useTableList 消费此结构） */
export interface PageResult<T> {
  list: T[]
  total: number
}

/**
 * 将任意后端分页包裹归一化为 PageResult<T>。
 * 兼容 { data, pagination: { total } } / { list, total } / { items, ... } 三种形态。
 */
export function toPageResult<T>(raw: unknown): PageResult<T> {
  if (raw == null) return { list: [], total: 0 }
  const r = raw as Record<string, unknown>
  const rawList = r.data ?? r.list ?? r.items
  const list = Array.isArray(rawList) ? (rawList as T[]) : []
  const pagination = (r.pagination ?? {}) as Partial<PaginationMeta>
  const total = Number(pagination.total ?? r.total ?? list.length) || 0
  return { list, total }
}
